plugins/alerting/pkg/alerting/api_logs.go

```diff
@@ -0,0 +1,176 @@
+			found = true
+			b = index
+			break
+		}
+	}
+	if found { // append to most recent bucket in existing index
+		if err := b.Append(event); err != nil {
+			return nil, err
+		}
+	} else {
+		if b == nil {
+			b = &BucketInfo{
+				ConditionId: event.ConditionId.Id,
+			}
+		}
+		if err := b.Create(); err != nil {
+			return nil, err
+		}
+		if err := b.Append(event); err != nil {
+			return nil, err
+		}
+	}
+	return &empty.Empty{}, nil
+}
+
+func (p *Plugin) GetAlertLog(ctx context.Context, ref *corev1.Reference) (*corev1.AlertLog, error) {
+	return nil, shared.AlertingErrNotImplemented
+}
+
+// https://stackoverflow.com/questions/18879109/subset-check-with-slices-in-go
+func isSubset(first, second []string) bool {
+	set := make(map[string]int)
+	for _, value := range second {
+		set[value] += 1
+	}
+
+	for _, value := range first {
+		if count, found := set[value]; !found {
+			return false
+		} else if count < 1 {
+			return false
+		} else {
+			set[value] = count - 1
+		}
+	}
+	return true
+}
+
+func hasLabels(item *alertingv1alpha.InformativeAlertLog, labels []string) bool {
+	return len(labels) == 0 || isSubset(labels, item.Condition.Labels)
+}
+
+func hasRange(item *alertingv1alpha.InformativeAlertLog, s *timestamppb.Timestamp, e *timestamppb.Timestamp) bool {
+	if item.Log.Timestamp == nil { // shouldn't happen, but hey
+		return true
+	}
+	start := s.AsTime()
+	end := s.AsTime()
+	if s == nil && e == nil {
+		return true
+	} else if s == nil { // before end
+		t := item.Log.Timestamp.AsTime()
+		return t.Before(end)
+	} else if e == nil {
+		t := item.Log.Timestamp.AsTime()
+		return t.After(start)
+	} else {
+		t := item.Log.Timestamp.AsTime()
+		return t.After(start) && t.Before(end)
+	}
+}
+
+func (p *Plugin) ListAlertLogs(ctx context.Context, req *alertingv1alpha.ListAlertLogRequest) (*alertingv1alpha.InformativeAlertLogList, error) {
+	if p.inMemCache != nil {
+		conditionsFrequentlyFired := make([]*corev1.AlertLog, 0)
+		keys := p.inMemCache.Keys()
+		for _, key := range keys {
+			val, _ := p.inMemCache.Peek(key)
+			conditionsFrequentlyFired = append(conditionsFrequentlyFired, val.(*corev1.AlertLog))
+		}
+		toBeFiltered := []*alertingv1alpha.InformativeAlertLog{}
+		for _, log := range conditionsFrequentlyFired {
+			var cnd *alertingv1alpha.AlertCondition
+			var err error
+			cnd, err = p.GetAlertCondition(ctx, log.ConditionId)
+			if err != nil {
+				// template an alert condition
+				cnd = &alertingv1alpha.AlertCondition{
+					Name:        "<not found>",
+					Description: "<not found>",
+				}
+			}
+			toBeFiltered = append(toBeFiltered, &alertingv1alpha.InformativeAlertLog{
+				Condition: cnd,
+				Log:       log,
+			})
+		}
+		toBeReturned := []*alertingv1alpha.InformativeAlertLog{}
+		for _, item := range toBeFiltered {
+			passed := true
+			passed = passed && hasLabels(item, req.Labels)
+			passed = passed && hasRange(item, req.StartTimestamp, req.EndTimestamp)
+			if passed {
```

Comment:
Can this be simplified to `if hasLabels(...) && hasRange(...) {` ?

Reply:
Yep